### ISSUE TITLE
Osimef849: Behavioural-Strategy & Implementation, 3 Weeks

### ISSUE THREAD

**Issue by Osimef849** [HackYourFutureBelgium/class-15#228]
[module repository](https://github.com/Osimef849/behavior-strategy-implementation.git)

# MY STUDY PLAN  & PRACTICE

- [ ] 🥚 **[./about-testing](./about-testing)**: A quick introduction to `describe`, `it`, and `expect`
- [ ] 🥚 **[./array-methods](./array-methods)**: Learn array methods by first understanding the strategies they are designed for.
  
- [ ] 🐣 **[./hoisting](./hoisting)**: That famous thing about JS! It's not so bad once you get it, the key to understanding hoisting is to understand the program life-cycle: _creation phase_ and _execution phase_. This chapter is more abstract, hoisting is something "invisible" about JavaScript program execution. These exercises will help you see hoisting by exploring the debugger at different points in program execution.
 - [ ] 🐣 **[./parsons-problems](./parsons-problems)**: Explore behavior, strategy and implementation by reconstructing different solutions to coding challenges.
 - [ ] 🐣 **[./function-design](./function-design)**: Learn a structured approach to understanding and solving coding challenges
 - [ ] 🐥 **[./throw-and-catch](./throw-and-catch)**: Go deeper into errors and error handling by throwing and catching your own errors.
- 🐥: **More Practice**
  - [ ] [rolling-scopes-school](https://github.com/rolling-scopes-school/js-assignments)
  - [ ] [Functions and Arrays](https://github.com/ironhack-labs/lab-javascript-functions-and-arrays/)
  - [ ] [Native Array Methods](https://github.com/gSchool/native-array-methods-exercises/)
  - [clue](https://github.com/ironhack-labs/lab-javascript-clue/)
  - [JavaScript-Koans](https://github.com/HackYourFutureBelgium/JavaScript-Koans/tree/master/src) (1-4)
  - [fuzz](https://hackyourfuturebelgium.github.io/fuzz)
  - [KBPsystem777](https://github.com/KBPsystem777/JSexercises)
  - [ ] [JavaScript Questions](https://github.com/lydiahallie/javascript-questions/)
  - [ ] [If Hemingway Wrote JavaScript](https://github.com/janke-learning/if-hemingway)

> [module repository](_)

**Comment by Osimef849** [comment]
### Check-In

#### I Need Help With:

Previous, study on certain topics in js and debugging modules such as 
how to do user interaction
I am abit confused on how to create functions, one time we declared a function using const
another time we called a function using some pre-defined libraries. (hope i am saying this right 😕 )


#### What went well?

Sunday class was well with @Samir, understanding the basis in BSI.
Still so many things seem to be in a rush in my head, but it gets clearer gradually.
Took out time to go through sundays video.

#### What went less well?

Trying to read other peoples code as the practice in BSI (no time),
I spent more time going through every segment of the video from sunday class.

#### Lessons Learned

In a nutshell BSI; Behaviour is the study of how codes are written and how they are executed in the lines, strategy describes what happens in each line of the syntax( its a good practice not to mention the feature used for example if we applied a while or for loop it is preferable to say iterate rather than "used for loop/while loop".

### Sunday Prep Work

Keep looking through what we done so far and practicing gradually.
Making good reference to edabit.com

**Comment by samirm00** [comment]
```
/*
There are three ways to create functions in JavaScript :

1- function declaration 

2- function expression 

3- arrow function (most used).

Note : there are some difference between function declaration and (function expression and arrow function)

like :

 Hoisting you will study it later in this module. (don't worry about it for now )

 `this` which is a keyword in JavaScript (don't worry about it for now )
*/

// 1- function declaration

function sayName(name) {
    console.log(`Hello ${name}`);
}

// 2- function expression it doesn't have a name that is why we assign it to a variable

const sayName = function (name) {
    console.log(`Hello ${name}`);
};

// 3- arrow function , the most used , because the syntax is very clear and very easy to use

const sayHello = (name) => {
    console.log(`Hello ${name}`);
};


// as for the functions you used from libraries or build in functions in JavaScript like : console.log() ,
// string.repeat()  , we call those methods 

```

**Comment by samirm00** [comment]
```
/*
As for Behaviour-strategy-implementation , I think you already understand well!
It it just you description is not quite accurate which is normal , you need sometime and practice to
be able to do  it accurately 

//Behaviour : it what function do , parameter and return values 

// Strategy : is just like a plan to write the function 

// implementation is just coding / writing your function 

Good Job!

Best of luck


*/
```

**Comment by Osimef849** [comment]
@samirm00  thanks this is a quick read for me. Honestly, HYF is intensive. Reading through so much information n studies gets me scared. But with this quick note reference, I will be able to make the study. Thank you.

**Comment by colevandersWands** [comment]
> I am abit confused on how to create functions

@Osimef849, take a look through the chapters in debugging/isolate about functions.  Those cover what you need to know for HYF.
